From ee74cb2a5ccdc13e8bf137d7387f01c6b202c150 Mon Sep 17 00:00:00 2001
From: Carlos Goncalves <carlos.goncalves@neclab.eu>
Date: Tue, 24 Jan 2017 21:52:27 +0000
Subject: [PATCH] API definition and reference for data plane status extension

Related-Bug: #1598081
Related-Bug: #1575146

Partial-Implements: blueprint port-data-plane-status

Change-Id: I04eef902b3310f799b1ce7ea44ed7cf77c74da04
---
 neutron_lib/api/definitions/base.py              |  1 +
 neutron_lib/api/definitions/data_plane_status.py | 78 ++++++++++++++++++++++++
 2 files changed, 79 insertions(+)
 create mode 100644 neutron_lib/api/definitions/data_plane_status.py

diff --git a/neutron_lib/api/definitions/base.py b/neutron_lib/api/definitions/base.py
index 6fbcbfa..38f183c 100644
--- a/neutron_lib/api/definitions/base.py
+++ b/neutron_lib/api/definitions/base.py
@@ -44,6 +44,7 @@ KNOWN_EXTENSIONS = (
     'auto-allocated-topology',
     'availability_zone',
     'binding',
+    'data-plane-status',
     'default-subnetpools',
     'dhcp_agent_scheduler',
     'dns-integration',
diff --git a/neutron_lib/api/definitions/data_plane_status.py b/neutron_lib/api/definitions/data_plane_status.py
new file mode 100644
index 0000000..5bcbf60
--- /dev/null
+++ b/neutron_lib/api/definitions/data_plane_status.py
@@ -0,0 +1,78 @@
+# Copyright (c) 2017 NEC Corporation.  All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from neutron_lib import constants
+
+
+VALID_VALUES = [None, constants.ACTIVE, constants.DOWN]
+
+# The alias of the extension.
+ALIAS = 'data-plane-status'
+
+# Whether or not this extension is simply signaling behavior to the user
+# or it actively modifies the attribute map.
+IS_SHIM_EXTENSION = False
+
+# Whether the extension is marking the adoption of standardattr model for
+# legacy resources, or introducing new standardattr attributes. False or
+# None if the standardattr model is adopted since the introduction of
+# resource extension.
+# If this is True, the alias for the extension should be prefixed with
+# 'standard-attr-'.
+IS_STANDARD_ATTR_EXTENSION = False
+
+# The name of the extension.
+NAME = 'Port data plane status extension'
+
+# The description of the extension.
+DESCRIPTION = "Expose status of underlying data plane"
+
+# A timestamp of when the extension was introduced.
+UPDATED_TIMESTAMP = "2017-01-24T10:00:00-00:00"
+
+# The name of the resource introduced or being extended.
+RESOURCE_NAME = 'port'
+
+# The plural for the resource introduced or being extended.
+COLLECTION_NAME = 'ports'
+
+# The specific resources and/or attributes for the extension (optional).
+DATA_PLANE_STATUS = 'data_plane_status'
+
+# The resource attribute map for the extension.
+RESOURCE_ATTRIBUTE_MAP = {
+    COLLECTION_NAME: {
+        DATA_PLANE_STATUS: {'allow_post': False, 'allow_put': True,
+                            'default': constants.ATTR_NOT_SPECIFIED,
+                            'validate': {'type:values': VALID_VALUES},
+                            'is_visible': True,
+                            'enforce_policy': True, }
+    },
+}
+
+# The subresource attribute map for the extension.
+SUB_RESOURCE_ATTRIBUTE_MAP = {
+}
+
+# The action map.
+ACTION_MAP = {
+}
+
+# The list of required extensions.
+REQUIRED_EXTENSIONS = [
+]
+
+# The list of optional extensions.
+OPTIONAL_EXTENSIONS = [
+]
-- 
2.12.3

